hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,197 @@
+{
+//! Spherical droplet geometry
+/*!
+ * TODO: write proper description of what is being done in the class below
+ */
+class __attribute__((visibility("default"))) SphereGeometry
+    {
+    public:
+        //! Constructor
+        /*!
+         * \param R confinement radius
+         * \param bc Boundary condition at the wall (slip or no-slip)
+         */
+        HOSTDEVICE SphereGeometry(Scalar R, boundary bc)
+            : m_R(R), m_R2(R*R), m_bc(bc)
+            { }
+
+        //! Detect collision between the particle and the boundary
+        /*!
+         * \param pos Proposed particle position
+         * \param vel Proposed particle velocity
+         * \param dt Integration time remaining
+         *
+         * \returns True if a collision occurred, and false otherwise
+         *
+         * \post The particle position \a pos is moved to the point of reflection, the velocity \a vel is updated
+         *       according to the appropriate bounce back rule, and the integration time \a dt is decreased to the
+         *       amount of time remaining.
+         */
+        HOSTDEVICE bool detectCollision(Scalar3& pos, Scalar3& vel, Scalar& dt) const
+            {
+
+            /*
+             * Detect if particle has left the spherical confinement.
+             * comparison: 'r2 <= R*R' used in the following calculations is equal to 0(FALSE) if the particle is outside
+             * a sphere of radius R, 1(TRUE) if inside.
+             */
+            Scalar r2 = dot(pos,pos);
+            // exit immediately if no collision is found.
+            /*
+            To avoid division by zero error later on, we exit immediately. (no collision could have occurred
+            if the particle speed is equal to zero)
+            */
+            Scalar v2 = dot(vel,vel);
+            if (r2 <= R2 || v2 == Scalar(0))
+               {
+               dt = Scalar(0);
+               return false;
+               }
+
+            // Find the point of contact when the particle is just leaving the surface
+            /*
+            * Calculating the time spent outside-bounds requires the knowledge of the point of contact on the boundary.
+            * The point of contact can be calculated using geometrical considerations.
+            *
+            * We know the following quantities,
+            *    1. r(t+del_t)      (the point outside the sphere)
+            *    2. v(t)            (previous velocity)
+            *
+            * Assuming 'r*' is the point of contact on the spherical shell and dt is the time particle travelled
+            * outside the boundary.
+            *
+            * r* = r(t+del_t) - dx* * v(t)/|v|
+            * and |r*|**2 = R**2
+            *
+            * Solving the above equations,
+            * dx* = (r.v^) - sqrt((r.v^)**2 - r2 + R2)
+            * consequently, dt = dx* / |v|
+            *
+            * --> dt = ((r.v) - sqrt((r.v)**2 - v2*(r2-R2)))/v2
+            */
+
```

Comment:
Condense this comment, the formula below is already clear:
```suggestion
            /*
            * Find the time remaining when the particle collided with the sphere of radius R. This time is
            * found by backtracking the position, r* = r-dt*v, and solving for dt when dot(r*,r*) = R^2.
            * This gives a quadratic equation in dt; the smaller root is the solution.
            */
```